core/src/com/unciv/ui/MultiplayerScreen.kt

```diff
@@ -0,0 +1,372 @@
+                    unselectGame()
+                }
+        ).pad(10f).row()
+
+        rightSideTable.add(
+                addGameButton.onClick {
+                    addMultiplayerGame(Gdx.app.clipboard.contents)
+                }
+        ).pad(10f).padBottom(30f).row()
+
+        rightSideTable.add(
+                refreshButton.onClick {
+                    redownloadAllGames()
+                }
+        ).pad(10f).row()
+
+        //RightSideButton Setup
+        rightSideButton.setText("Join Game".tr())
+        rightSideButton.onClick {
+            joinMultiplaerGame()
+        }
+    }
+
+    //Adds a new Multiplayer game to the List
+    fun addMultiplayerGame(gameId: String, gameName: String = ""){
+        if (gameIsAlreadySavedAsMultiplayer(gameId)) {
+            ResponsePopup("Game is already added".tr(), this)
+            return
+        }
+        try {
+            //since the gameId is a String it can contain anything and has to be checked
+            UUID.fromString(gameId.trim())
+        } catch (ex: Exception) {
+            Popup(this).apply {
+                addGoodSizedLabel("Invalid game ID!".tr())
+                row()
+                addCloseButton()
+                open()
+            }
+            return
+        }
+        thread(name="MultiplayerDownload") {
+            addGameButton.setText("Working...".tr())
+            addGameButton.disable()
+            try {
+                // The tryDownload can take more than 500ms. Therefore, to avoid ANRs,
+                // we need to run it in a different thread.
+                val game = OnlineMultiplayer().tryDownloadGame(gameId.trim())
+                if (gameName == "")
+                    GameSaver().saveGame(game, game.gameId, true)
+                else
+                    GameSaver().saveGame(game, gameName, true)
+
+                reloadGameListUI()
+            } catch (ex: Exception) {
+                Popup(this).apply {
+                    addGoodSizedLabel("Could not download game!".tr())
+                    row()
+                    addCloseButton()
+                    open()
+                }
+            }
+            addGameButton.setText(addGameText)
+            addGameButton.enable()
+        }
+    }
+
+    //just loads the game from savefile
+    //the game will be downloaded opon joining it anyway
+    private fun joinMultiplaerGame(){
+        try {
+            UncivGame.Current.loadGame(selectedGame)
+        } catch (ex: Exception) {
+            Popup(this).apply {
+                addGoodSizedLabel("Could not download game!".tr())
+                row()
+                addCloseButton()
+                open()
+            }
+            return
+        }
+    }
+
+    private fun gameIsAlreadySavedAsMultiplayer(gameId: String) : Boolean{
+        return multiplayerGameList.containsKey(gameId)
+    }
+
+    //reloads all gameFiles to refresh UI
+    fun reloadGameListUI(){
+        try {
```

Comment:
Why do we have 2 'try's? What errors does the first one try to catch?

Reply:
If the first try catches, the whole reload process is scraped. It catches the gameSaver.getSaves(). The second try is for catching a specific game reload. So the reload can still happen for every other game.
The inner one is not super necessary, but I thought it could be useful.

Reply:
Sounds good, but for readability let's have it sequential and not nested.
Try{...saves=getsaves()}catch{...return}
For  (save in saves) try{...} Catch{...}